### ISSUE TITLE
Memory leak in consecutive YOLO training sessions

### ISSUE THREAD

**Issue by hyz19961004** [ultralytics/ultralytics#21490]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Description
When running multiple consecutive YOLO training sessions in the same Python process, there is a significant memory leak. The RSS memory usage grows from ~500MB to over 3.3GB after just 3 training sessions and never gets released, even after the trainer objects are deleted. Or do I need to do some additional manual memory recovery?

### Environment
- **Ultralytics Version**: 8.3.156
- **Python Version**: 3.9.5  
- **PyTorch Version**: 2.7.1+cu128
- **CUDA Version**: CUDA:0 (NVIDIA GeForce RTX 4060, 8188MiB)
- **OS**: Linux 5.15.0-139-generic

### Problem Analysis
**Memory Growth Pattern:**
- **1st training session**: 558MB → 3,026MB (+2,468MB)
- **2nd training session**: 3,114MB → 3,260MB (+146MB cumulative)  
- **3rd training session**: 3,260MB → 3,418MB (+158MB cumulative)

**Expected behavior**: Memory should return to baseline (~500-600MB) after each training session.
**Actual behavior**: Memory accumulates and is never released, causing issues in long-running applications.

### Reproduction Code

```python
from ultralytics import YOLO
from datetime import datetime
from pathlib import Path
from typing import Tuple, Optional, Dict
import yaml
import tracemalloc
import psutil
import time
import threading


class ModelTrain:
    def __init__(self):
        take_memory_snapshot("ModelTrain Init")

    def set_config(self, work_project: Path, config: Path, model_path: str):
        self.project = work_project
        self.config_path = config
        self.model_path = model_path

    def train(self, batch: Optional[int] = None) -> Tuple[str, str, str, Dict]:
        try:
            model = YOLO(self.model_path, task='detect')
            take_memory_snapshot("train start")
            
            # Training with standard parameters
            model.train(
                data=self.config_path,
                project=self.project,
                name='train',
                epochs=3,
                patience=30,
                imgsz=640,
                batch=batch,
                device=0,
                workers=4,
                save=True,
                cache=False,
                rect=True,
                # ... other parameters
            )
            take_memory_snapshot("train end")

            # Validation
            val_results = model.val(
                data=self.config_path,
                split='test',
                device=0,
                imgsz=1280,
                batch=batch,
                rect=True,
                project=self.project,
                name='validation',
                iou=0.4,
            )
            take_memory_snapshot("val end")
            
            # Process results...
            return recall, fpr, mAP50, class_metrics

        except Exception as e:
            print(f"Training error: {str(e)}")


def take_memory_snapshot(description: str):
    try:
        process = psutil.Process()
        memory_info = process.memory_info()
        system_memory = psutil.virtual_memory()
        
        rss_mb = memory_info.rss / 1024 / 1024
        system_available_gb = system_memory.available / 1024 / 1024 / 1024
        
        print(f"[{description}]: RSS={rss_mb:.1f}MB, "
              f"system_available={system_available_gb:.1f}GB, "
              f"system_used={system_memory.percent:.1f}%")
    except Exception as e:
        print(f"Memory snapshot error: {str(e)}")


def reproduce_memory_leak():
    """Reproduce the memory leak with 3 consecutive training sessions"""
    tracemalloc.start()
    
    for i in range(3):
        print(f"\n=== Training Session {i+1} ===")
        trainer = ModelTrain()
        trainer.set_config(
            work_project=Path("./output/"),
            config=Path("./data.yaml"), 
            model_path="yolov8s.pt"
        )
        trainer.train(batch=4)
        
        # Even explicit deletion doesn't help
        del trainer


if __name__ == "__main__":
    reproduce_memory_leak()
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @hyz19961004, thank you for reporting this and for the detailed information! 🚀 This is an automated response to help you get the fastest support possible. An Ultralytics engineer will also review your issue and assist you soon.

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for additional resources, including many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common issues are addressed there.

Since this appears to be a 🐛 Bug Report, please make sure you’ve provided a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/). The code snippet you shared is very helpful, but if possible, please further minimize it to only the lines required to reproduce the memory leak. This will help our team debug more efficiently.

If you have additional logs, environment details, or can reproduce with default configs or models, that would also be helpful!

Join the Ultralytics community for real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧, in-depth discussions on [Discourse](https://community.ultralytics.com/), or connect with others on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). Upgrading may resolve your issue if it has already been addressed in a recent release:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for helping us improve Ultralytics! 🧠

**Comment by glenn-jocher** [comment]
This is a known issue with PyTorch/CUDA memory management in consecutive training sessions. The memory accumulation occurs because GPU tensors and model states aren't fully released between sessions.

Try adding explicit memory cleanup between training sessions:

```python
import torch
import gc

def reproduce_memory_leak():
    for i in range(3):
        print(f"\n=== Training Session {i+1} ===")
        trainer = ModelTrain()
        trainer.set_config(
            work_project=Path("./output/"),
            config=Path("./data.yaml"), 
            model_path="yolov8s.pt"
        )
        trainer.train(batch=4)
        
        # Explicit cleanup
        del trainer
        gc.collect()
        torch.cuda.empty_cache()
        time.sleep(1)  # Allow cleanup to complete
```

For production environments with multiple training sessions, consider running each session in a separate subprocess to ensure complete memory isolation, as PyTorch's memory management can retain references that prevent full cleanup within the same process.

**Comment by hyz19961004** [comment]
> This is a known issue with PyTorch/CUDA memory management in consecutive training sessions. The memory accumulation occurs because GPU tensors and model states aren't fully released between sessions.
> 
> Try adding explicit memory cleanup between training sessions:
> 
> import torch
> import gc
> 
> def reproduce_memory_leak():
>     for i in range(3):
>         print(f"\n=== Training Session {i+1} ===")
>         trainer = ModelTrain()
>         trainer.set_config(
>             work_project=Path("./output/"),
>             config=Path("./data.yaml"), 
>             model_path="yolov8s.pt"
>         )
>         trainer.train(batch=4)
>         
>         # Explicit cleanup
>         del trainer
>         gc.collect()
>         torch.cuda.empty_cache()
>         time.sleep(1)  # Allow cleanup to complete
> For production environments with multiple training sessions, consider running each session in a separate subprocess to ensure complete memory isolation, as PyTorch's memory management can retain references that prevent full cleanup within the same process.

Thank you for your reply. The problem was not solved after I tried to recycle manually. I will try to move the training module to the subprocess.

**Comment by glenn-jocher** [comment]
You're right that the basic cleanup approach often isn't sufficient for complete memory recovery in PyTorch. The subprocess approach is indeed the most reliable solution for this issue.

Additionally, you might want to try using the `cuda_memory_usage` context manager from Ultralytics for better memory monitoring during training:

```python
from ultralytics.utils.torch_utils import cuda_memory_usage

def train_with_monitoring(self, batch: Optional[int] = None):
    with cuda_memory_usage(device=0) as cuda_info:
        model = YOLO(self.model_path, task='detect')
        model.train(...)
    print(f"CUDA memory used: {cuda_info['memory']:.2f}GB")
```

The subprocess approach will completely isolate memory between sessions and prevent the accumulation you're experiencing. This is the recommended solution for production environments with consecutive training runs.
